File: csharp/src/detectors/hardcoded-credentials/HardcodedCredentials.cs
using System.Net;

namespace Credentials
{
    class HardcodedCredentials
    {
        // {fact rule=hardcoded-credentials@v1.0 defects=1}
        public void HardcodedCredentialsNoncompliant()
        {
            string data;
            data = "7e5tc4s3";
            if (data != null)
            {
                // Noncompliant: Hardcoded data used as password in `NetworkCredential()`.
                NetworkCredential credentials = new NetworkCredential("user", data, "domain");
                Console.WriteLine(credentials.ToString());
            }
        }
        // {/fact}
        
        // {fact rule=hardcoded-credentials@v1.0 defects=0}
        public void HardcodedCredentialsCompliant()
        {
            string data;
            data = "";
            // Compliant: Read data from the console using `ReadLine()` used it in `NetworkCredential()`.
            try
            {
                data = Console.ReadLine();
            }
            catch (IOException exceptIO)
            {
                Console.WriteLine($"Error with stream reading {exceptIO}");
            }
            if (data != null)
            {
                NetworkCredential credentials = new NetworkCredential("user", data, "domain");
                Console.WriteLine(credentials.ToString());
            }
        }
        // {/fact}
    }
}

File: typescript/src/detector/critical/hardcoded_credentials/hardcoded_credentials.ts
// {fact rule=hardcoded-credentials@v1.0 defects=1}

function nonCompliant() {
  var crypto = require("crypto");
  function hardcodedCredentialsNoncompliant(
    salt: any,
    iterations: any,
    keyLen: any,
    digest: any,
  ) {
    // Noncompliant: password is hardcoded.
    crypto.pbkdf2(
      "password",
      salt,
      iterations,
      keyLen,
      digest,
      (err: any, key: { toString: (arg0: string) => any }) => {
        if (err) {
          throw err;
        }
        return key.toString("base64");
      },
    );
  }
}
// {/fact}

// {fact rule=hardcoded-credentials@v1.0 defects=0}

function compliant() {
  var crypto = require("crypto");
  function hardcodedCredentialsCompliant(
    salt: any,
    iterations: any,
    keyLen: any,
    digest: any,
  ) {
    // Compliant: password is obtained from environment.
    crypto.pbkdf2(
      process.env.password,
      salt,
      iterations,
      keyLen,
      digest,
      (err: any, key: { toString: (arg0: string) => any }) => {
        if (err) {
          throw err;
        }
        return key.toString("base64");
      },
    );
  }
}
// {/fact}
